### PR TITLE
KAFKA-15854: Move Java classes from `kafka.server` to the `server` module

### PR DESCRIPTION
**Pull request by ijuma**
We only move Java classes that have minimal or no dependencies on Scala classes in this PR.

Details:
* Configured `server` module in build files.
* Changed `ControllerRequestCompletionHandler` to be an interface since it has no implementations.
* Cleaned up various import control files.
* Minor build clean-ups for `server-common`.
* Disabled `testAssignmentAggregation` when executed with Java 8, this is an existing issue (see #14794).

For broader context on this change, please check:
* KAFKA-15852: Move server code from `core` to `server` module

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -1643,19 +1700,6 @@ project(':server-common') {
...
-      java {
-        srcDirs = ["src/main/java"]
-      }
-    }
-    test {
-      java {
-        srcDirs = ["src/test/java"]
-      }
-    }
-  }
```
Comment (ijuma) [inline build.gradle:1657]: Redundant.

File: build.gradle [hunk 2]
```diff
@@ -1670,6 +1714,10 @@ project(':server-common') {
   checkstyle {
     configProperties = checkstyleConfigProperties("import-control-server-common.xml")
   }
+
+  javadoc {
+    enabled = false
```
Comment (ijuma) [inline build.gradle:1719]: We don't have public classes in `server-common`, so make it explicit that javadoc should not be published.

File: server/src/main/java/org/apache/kafka/server/metrics/ClientMetricsConfigs.java [hunk 3]
```diff
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package kafka.metrics;
+package org.apache.kafka.server.metrics;
```
Comment (ijuma) [inline server/src/main/java/org/apache/kafka/server/metrics/ClientMetricsConfigs.java:17]: @junrao @apoorvmittal10 Note that I have moved these recently introduced classes to this new module.

File: build.gradle [hunk 4]
```diff
@@ -842,6 +842,62 @@ tasks.create(name: "jarConnect", dependsOn: connectPkgs.collect { it + ":jar" })
...
+
+    testImplementation project(':clients').sourceSets.test.output
+
+    testImplementation libs.mockitoCore
+    testImplementation libs.junitJupiter
+    testImplementation libs.slf4jlog4j
+  }
+
+  task createVersionFile() {
+    def receiptFile = file("$buildDir/kafka/$buildVersionFileName")
```
Comment (divijvaidya) [inline build.gradle:864]: am not a gradle expert but my IDE is telling me that $buildDir is deprecated and replaced by `getLayout().getBuildDirectory()` instead

File: build.gradle [hunk 5]
```diff
@@ -842,6 +842,62 @@ tasks.create(name: "jarConnect", dependsOn: connectPkgs.collect { it + ":jar" })
...
+    compileOnly libs.log4j
+
+    testImplementation project(':clients').sourceSets.test.output
+
+    testImplementation libs.mockitoCore
+    testImplementation libs.junitJupiter
+    testImplementation libs.slf4jlog4j
+  }
+
+  task createVersionFile() {
```
Comment (divijvaidya) [inline build.gradle:863]: this is duplicated across all projects. Could we extract it out and use the same task across all sub-projects?

File: checkstyle/import-control-storage.xml [hunk 6]
```diff
@@ -27,15 +27,11 @@
...
     <allow pkg="org.junit" />
-    <allow pkg="org.opentest4j" />
     <allow pkg="org.hamcrest" />
     <allow pkg="org.mockito" />
-    <allow pkg="org.easymock" />
-    <allow pkg="org.powermock" />
     <allow pkg="java.security" />
     <allow pkg="javax.net.ssl" />
     <allow pkg="javax.security" />
-    <allow pkg="org.ietf.jgss" />
```
Comment (divijvaidya) [inline checkstyle/import-control-storage.xml:38]: thank you for cleaning the import-control files!

File: build.gradle (hunk 1 above)
Comment (divijvaidya) [reply build.gradle:1657]: Should we create a ticket to get rid of this from all sub-projects? If you like, we can do it as part of this PR itself.

**Review by divijvaidya — COMMENTED** [review]
Please add a package-info.java file in the server which can help new contributors to the code understand what components goes in the `server` module. At some stage we probably need a component level diagram of Kafka code base to improve onboarding to code base.

File: build.gradle (hunk 4 above)
Comment (ijuma) [reply build.gradle:864]: I used the same pattern used by other modules. We can replace all of them via a separate PR.

File: build.gradle (hunk 5 above)
Comment (ijuma) [reply build.gradle:863]: I'm not sure. Best done via its own PR as well.

File: build.gradle (hunk 1 above)
Comment (ijuma) [reply build.gradle:1657]: I can file a ticket. Some of them are actually necessary as they add additional directories.

File: build.gradle (hunk 4 above)
Comment (divijvaidya) [reply build.gradle:864]: ok

File: build.gradle (hunk 5 above)
Comment (divijvaidya) [reply build.gradle:863]: ok

File: server/src/main/java/org/apache/kafka/server/metrics/ClientMetricsConfigs.java (hunk 3 above)
Comment (apoorvmittal10) [reply server/src/main/java/org/apache/kafka/server/metrics/ClientMetricsConfigs.java:17]: Thanks @ijuma. @junrao I ll move the classes in other PRs to server/src package itself once review is complete. Moving prior active review will lose context of comments.

File: build.gradle (hunk 4 above)
Comment (ijuma) [reply build.gradle:864]: Fyi :

"Note that, at this stage, Gradle will not print deprecation warnings if you still use Project.buildDir. We know this is a big change and want to give time for authors of major plugins to move away from its usage first."

**Comment by ijuma** [comment]
@divijvaidya Thanks for the review. I added a basic `package-info` file. We can flesh it out as we add move code to this module.

File: server/src/main/java/org/apache/kafka/server/metrics/ClientMetricsConfigs.java (hunk 3 above)
Comment (ijuma) [reply server/src/main/java/org/apache/kafka/server/metrics/ClientMetricsConfigs.java:17]: Can you please share the PRs where this is being discussed now?

**Review by divijvaidya — APPROVED** [review]

**Comment by divijvaidya** [comment]
Please feel free to merge after CI is successful.

**Comment by ijuma** [comment]
I ran the tests a few times and the JDK 8 build passed, the failures for the other builds are unrelated:

> Build / JDK 17 and Scala 2.13 / org.apache.kafka.trogdor.coordinator.CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated()
Build / JDK 21 and Scala 2.13 / org.apache.kafka.connect.integration.ConnectorRestartApiIntegrationTest.testMultiWorkerRestartOnlyConnector
Build / JDK 21 and Scala 2.13 / kafka.api.DelegationTokenEndToEndAuthorizationWithOwnerTest.testProduceConsumeViaAssign(String).quorum=kraft
Build / JDK 21 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerMetricsIntegrationTest.testTimeoutMetrics()
Build / JDK 21 and Scala 2.13 / org.apache.kafka.tiered.storage.integration.TransactionsWithTieredStoreTest.testBumpTransactionalEpoch(String).quorum=kraft
Build / JDK 11 and Scala 2.13 / kafka.api.SaslPlainPlaintextConsumerTest.testCoordinatorFailover()

**Pull request merged**